kernel/libs/jhash/src/lib.rs

```diff
@@ -0,0 +1,331 @@
+    )
+}
+
+/// Hashes exactly one u32 value
+pub const fn jhash_1vals(a: u32, initval: u32) -> u32 {
+    jhash_3vals_inner(
+        a,
+        0,
+        0,
+        initval.wrapping_add(JHASH_INITVAL).wrapping_add(1 << 2),
+    )
+}
+
+/// Hashes an array of u32 values.
+pub const fn jhash_u32_array(array: &[u32], initval: u32) -> u32 {
+    let mut length = array.len() as u32;
+    let mut index = 0;
+
+    // Initialize values a, b, and c
+    let mut a: u32 = JHASH_INITVAL
+        .wrapping_add(length << 2)
+        .wrapping_add(initval);
+    let mut b: u32 = a;
+    let mut c: u32 = a;
+
+    // Process most values except the last three
+    while length > 3 {
+        a = a.wrapping_add(array[index]);
+        b = b.wrapping_add(array[index + 1]);
+        c = c.wrapping_add(array[index + 2]);
+
+        (a, b, c) = jhash_mix(a, b, c);
+        length -= 3;
+        index += 3;
+    }
+
+    // Handle the last three values
+    debug_assert!(length <= 3);
+
+    if length == 3 {
+        c = c.wrapping_add(array[index + 2]);
+    }
+
+    if length >= 2 {
+        b = b.wrapping_add(array[index + 1]);
+    }
+
+    if length >= 1 {
+        a = a.wrapping_add(array[index]);
+        return jhash_final(a, b, c);
+    }
+
+    c
+}
+
+/// An internal function that handles hashing for 3 u32 values
+const fn jhash_3vals_inner(mut a: u32, mut b: u32, mut c: u32, initval: u32) -> u32 {
+    a = a.wrapping_add(initval);
+    b = b.wrapping_add(initval);
+    c = c.wrapping_add(initval);
+
+    jhash_final(a, b, c)
+}
+
+/// Finalizes the mix of three 32-bit values into a single u32 value
+const fn jhash_final(mut a: u32, mut b: u32, mut c: u32) -> u32 {
+    c ^= b;
+    c = c.wrapping_sub(b.rotate_left(14));
+
+    a ^= c;
+    a = a.wrapping_sub(c.rotate_left(11));
+
+    b ^= a;
+    b = b.wrapping_sub(a.rotate_left(25));
+
+    c ^= b;
+    c = c.wrapping_sub(b.rotate_left(16));
+
+    a ^= c;
+    a = a.wrapping_sub(c.rotate_left(4));
+
+    b ^= a;
+    b = b.wrapping_sub(a.rotate_left(14));
+
+    c ^= b;
+    c.wrapping_sub(b.rotate_left(24))
+}
+
+/// Mixes three 32-bit values in a reversible manner
+const fn jhash_mix(mut a: u32, mut b: u32, mut c: u32) -> (u32, u32, u32) {
+    a = a.wrapping_sub(c);
+    a ^= c.rotate_left(4);
+    c = c.wrapping_add(b);
+
+    b = b.wrapping_sub(a);
+    b ^= a.rotate_left(6);
+    a = a.wrapping_add(c);
+
+    c = c.wrapping_sub(b);
+    c ^= b.rotate_left(8);
+    b = b.wrapping_add(a);
+
+    a = a.wrapping_sub(c);
+    a ^= c.rotate_left(16);
+    c = c.wrapping_add(b);
+
+    b = b.wrapping_sub(a);
+    b ^= a.rotate_left(19);
+    a = a.wrapping_add(c);
+
+    c = c.wrapping_sub(b);
+    c ^= b.rotate_left(4);
+    b = b.wrapping_add(a);
+
+    (a, b, c)
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    const JHASH_INITVAL: u32 = 0;
+
+    #[test]
+    fn test_jhash_3vals() {
+        assert_eq!(jhash_3vals(1, 2, 3, JHASH_INITVAL), 2757843189);
```

Comment:
TBH, I don't know how to maintain these values. Do you run the Linux equivalent to make sure the values are correct?

Reply:
Yes, I agree such values look strange. But it seems we can't include it's C's counterpart to ensure they behave the same.

>  Do you run the Linux equivalent to make sure the values are correct?

Yes, I have wrote a C program(mostly copied from Linux, change some code to make it successfully compiled) to test the values. I only test part cases. And the results are all the same. I skip `jhash_slice` because I did not find a corresponding function for `__get_unaligned_cpu32`

<details><summary>Corresponding C Jhash program</summary>
<p>

```C
#include <stdio.h>
#include <stdint.h>

typedef uint8_t u8;
typedef uint32_t u32;
#define fallthrough __attribute__((fallthrough))

static inline u32 rol32(u32 word, unsigned int shift)
{
	return (word << (shift & 31)) | (word >> ((-shift) & 31));
}

/* __jhash_mix - mix 3 32-bit values reversibly. */
#define __jhash_mix(a, b, c)			\
{						\
	a -= c;  a ^= rol32(c, 4);  c += b;	\
	b -= a;  b ^= rol32(a, 6);  a += c;	\
	c -= b;  c ^= rol32(b, 8);  b += a;	\
	a -= c;  a ^= rol32(c, 16); c += b;	\
	b -= a;  b ^= rol32(a, 19); a += c;	\
	c -= b;  c ^= rol32(b, 4);  b += a;	\
}

/* __jhash_final - final mixing of 3 32-bit values (a,b,c) into c */
#define __jhash_final(a, b, c)			\
{						\
	c ^= b; c -= rol32(b, 14);		\
	a ^= c; a -= rol32(c, 11);		\
	b ^= a; b -= rol32(a, 25);		\
	c ^= b; c -= rol32(b, 16);		\
	a ^= c; a -= rol32(c, 4);		\
	b ^= a; b -= rol32(a, 14);		\
	c ^= b; c -= rol32(b, 24);		\
}

/* An arbitrary initial parameter */
#define JHASH_INITVAL (uint32_t)0xdeadbeef

/* jhash2 - hash an array of u32's
 * @k: the key which must be an array of u32's
 * @length: the number of u32's in the key
 * @initval: the previous hash, or an arbitrary value
 *
 * Returns the hash value of the key.
 */
static inline u32 jhash2(const u32 *k, u32 length, u32 initval)
{
	u32 a, b, c;

	/* Set up the internal state */
	a = b = c = JHASH_INITVAL + (length<<2) + initval;

	/* Handle most of the key */
	while (length > 3) {
		a += k[0];
		b += k[1];
		c += k[2];
		__jhash_mix(a, b, c);
		length -= 3;
		k += 3;
	}

	/* Handle the last 3 u32's */
	switch (length) {
	case 3: c += k[2];	fallthrough;
	case 2: b += k[1];	fallthrough;
	case 1: a += k[0];
		__jhash_final(a, b, c);
		break;
	case 0:	/* Nothing left to add */
		break;
	}

	return c;
}


/* __jhash_nwords - hash exactly 3, 2 or 1 word(s) */
static inline u32 __jhash_nwords(u32 a, u32 b, u32 c, u32 initval)
{
	a += initval;
	b += initval;
	c += initval;

	__jhash_final(a, b, c);

	return c;
}

static inline u32 jhash_3words(u32 a, u32 b, u32 c, u32 initval)
{
	return __jhash_nwords(a, b, c, initval + JHASH_INITVAL + (3 << 2));
}

static inline u32 jhash_2words(u32 a, u32 b, u32 initval)
{
	return __jhash_nwords(a, b, 0, initval + JHASH_INITVAL + (2 << 2));
}

static inline u32 jhash_1word(u32 a, u32 initval)
{
    return __jhash_nwords(a, 0, 0, initval + JHASH_INITVAL + (1 << 2));
}

int main() {
    printf("%u\n", jhash_1word(1, 0));
    printf("%u\n", jhash_1word(5, 0));
    printf("%u\n", jhash_1word(5, 0));
    printf("%u\n", jhash_1word(0, 0));
    printf("%u\n", jhash_1word(0xffffffff, 0));
    printf("%u\n", jhash_1word(1, 10));
    printf("%u\n", jhash_1word(10, 5));

    printf("%u\n", jhash_2words(1, 2, 0));

    printf("%u\n", jhash_3words(1, 2, 3, 0));

    int a[3] = {1,2,3};
    printf("%u\n", jhash2(a, 3, 0));
    int b[4] = {0,1,2,0xffffffff};
    printf("%u\n", jhash2(b, 4, 0));
    return 0;
}
```

</p>
</details>

Reply:
Thanks, I'm asking because you seem to miss a `__jhash_mix` (https://github.com/asterinas/asterinas/pull/1750#discussion_r1898261343), but the tests did not help you find the bug.

> I skip `jhash_slice` because I did not find a corresponding function for `__get_unaligned_cpu32`

It's just `*(int *)xx`, I think.

> But it seems we can't include it's C's counterpart to ensure they behave the same.

Of course, I understand it's not possible for us to include the C program in our codebase.